### PR TITLE
Fix deduplication results displaying wrong data

### LINKED ISSUE #15588: Explore: Not cleaning logs when change of duplicate method
### What Grafana version are you using?
Grafana v6.0.0-pre3 (commit: cdac119)

### What datasource are you using?
Loki

### What OS are you running grafana on?
Docker image (:master)

### What did you do?
On `Dedup` option (in `None`) select another of the Options, ex. `Signature`

### What was the expected result?
That all logs is cleaned and grouped by signature.

### What happened instead?
All logs is grouped by signature, but no cleaned the old ones.

deepin-screen-recorder_select area_20190221132028

The problem is from `None` to another one.

### PR DESCRIPTION
**Pull request by dprokop**
Fixes #15588

Since the log results are changing in redux store when dedup strategy changes (`duplicates` field is added to log result), the memoization hash was not deterministic and wrong cached values for log results selector could be used.

### CODE REVIEW

**Comment by davkal** [comment]
@dprokop how can I repro the original issue?

What could help here is a test that fails without the fix, and passes with the fix.

**Comment by dprokop** [comment]
@davkal sorry for not providing more details. 

To reproduce the issue change dedup strategy in following sequence: none -> exact -> number -> signature couple of times. Yes, I know this is quite strange, but this is the only way I could repro this issue. Also, cannot repro in test even when performing on the same data set that it fails in the UI.

**Comment by davkal** [comment]
I'm 95% certain that #15588 is a react key issue. Log entries currently use the label+timestamp string as react keys. In the issue's video, all initial items probably have the same key. After deduping a new list comes in with just one deduped item, but it has *the same* key as the existing items, so those dont get unmounted, just the deduped row gets added. Worse, the deduped row has an additional column, so it all looks a bit offset.

I don't see how this PR solves this issue.

The custom selector in the existing code could use a comment on what problem it solves. It makes it quite difficult to follow what's happening with these higher order functions.
Would this custom selector have other issues?

**Comment by dprokop** [comment]
Well, that's an interesting observation David. Gonna dig deeper then, thanks!

**Review by torkelo — APPROVED** [review]
LGTM

**Comment by davkal** [comment]
The fix of this PR is for an issue in the dedup code itself, independent of the selector. I suspect it can be already be reproduced by running `dedupLogRows` twice on the same log result. Might be worth keeping the selector test as well though, but IMO it's too noisy and should be reduced to the minimal case that fails (before the fix).

**Comment by dprokop** [comment]
@davkal I have run the selector test on limited data set:
```
[
  {
    entry: 'WARN test 1.23 on [xxx]',
  },
  {
    entry: 'WARN test 1.24 on [xxx]',
  },
  {
    entry: 'WARN test 1.25 on [xxx]',
  }
]
```

Also, run a test of `dedupLogRows` on the same data set without the fix applied. Here are my findings:

1. `dedupLogRows` seems to be working fine when run multiple times on the same log results
2. I can reproduce the problem only when cycling through dedup strategies twice.
2. I still think the problem lies in the memoization of the dedup results, but I cannot fully understand following thing when running the test I have implemented (the noisy one). I will focus on `exact` strategy, as it seems like the problem surfaces only with it:
- First time we execute exact strategy the selector is run and the result is memoized under key `-{"rows":[{"entry":"WARN test 1.23 on [xxx]"},{"entry":"WARN test 1.24 on [xxx]"},{"entry":"WARN test 1.25 on [xxx]"}]}-undefined-"exact"` and the result is:
``` 
{ rows:
         [ { entry: 'WARN test 1.23 on [xxx]', duplicates: 0 },
           { entry: 'WARN test 1.24 on [xxx]', duplicates: 0 },
           { entry: 'WARN test 1.25 on [xxx]', duplicates: 0 } ] }

```
-
- Second time we execute exact strategy the selector is run and the result is memoized under key `-{"rows":[{"entry":"WARN test 1.23 on [xxx]","duplicates":2},{"entry":"WARN test 1.24 on [xxx]","duplicates":0},{"entry":"WARN test 1.25 on [xxx]","duplicates":0}]}-undefined-"exact"` and the result is:
```
{ rows:
         [ { entry: 'WARN test 1.23 on [xxx]', duplicates: 0 },
           { entry: 'WARN test 1.24 on [xxx]', duplicates: 0 },
           { entry: 'WARN test 1.25 on [xxx]', duplicates: 0 } ] }
```
- Third and following times we execute exact strategy the selector is no longer run, the memoization key is `-{"rows":[{"entry":"WARN test 1.23 on [xxx]","duplicates":2},{"entry":"WARN test 1.24 on [xxx]","duplicates":0},{"entry":"WARN test 1.25 on [xxx]","duplicates":0}]}-undefined-"exact"` (the same as for 2nd execution, but the result is 
```
{ rows:
         [ { entry: 'WARN test 1.23 on [xxx]', duplicates: 2 },
           { entry: 'WARN test 1.24 on [xxx]', duplicates: 0 },
           { entry: 'WARN test 1.25 on [xxx]', duplicates: 0 } ] }
```

For me it seems like it's an issue with the memoised result keeping a references to the log row from redux store, and since before the fix we were modifying the log rows in store when calculating deduped rows instead of making a copy of the row, the memoized result basically is not stable, as the referenced elements can be modified.

WDYT?

**Comment by davkal** [comment]
Since the function that the selector runs is modifying the input, we're in trouble. It will always have to run the selector function twice for what should be the same input.

File: public/app/features/explore/state/selectors.test.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import { deduplicatedLogsSelector } from './selectors';
+import { LogsDedupStrategy } from 'app/core/logs_model';
+import { ExploreItemState } from 'app/types';
+
+const state = {
+  logsResult: {
+    rows: [
+      {
+        entry: '2019-03-05T11:00:56Z sntpc sntpc[1]: offset=-0.033938, delay=0.000649',
```
Comment (davkal) [inline public/app/features/explore/state/selectors.test.ts]: I'd use text here that makes it more clear what characteristic of the line is important for the deduping.

**Pull request merged**